# MetaAML_CIBERSORTX_VAF_analysis
#
# Brooks Benard
# bbenard@stanford.edu
# 07/15/2019
#
# CYBERsort predictions generated by Asiri Ediriwickrema (asiri@stanford.edu)

### This script used CYBERsort-predicted cell type abundances and performs mutation-specific analysis, VAF regression on cell type abundances, drug sensitivity analysis, and clinical survival analysis on results from the Beat AML dataset.

#### load required packages and data ####
if (!require('ggplot2')) install.packages('ggplot2'); library('ggplot2')
if (!require('readxl')) install.packages('readxl'); library('readxl')
if (!require('cowplot')) install.packages('cowplot'); library('cowplot')
if (!require('reshape2')) install.packages('reshape2'); library('reshape2')
if (!require('plyr')) install.packages('plyr'); library('plyr')
if (!require('dplyr')) install.packages('dplyr'); library('dplyr')
if (!require('data.table')) install.packages('data.table'); library('data.table')
if (!require('ggrepel')) install.packages('ggrepel'); library('ggrepel')


# read in the CIBERsort results and do some reformatting to merge with vaf data
CIBERSORT_results <- read.csv("~/Desktop/Majeti_Lab/Data/CIBERSORTx/master_all.csv")
colnames(CIBERSORT_results)[1] <- "Sample"
CIBERSORT_results$Sample <- gsub(CIBERSORT_results$Sample, pattern = "\\.", replacement = "-")  

# load the MetaAML results file and annotate sample names to match the cibersort annotations
load("~/Desktop/MetaAML_results/final_data_matrix.RData")

for(i in 1:nrow(final_data_matrix)){
  if(!is.na(final_data_matrix$Cohort[i])){
    if(final_data_matrix$Cohort[i] == "Tyner"){
      final_data_matrix$Sample[i] <- final_data_matrix$LabId[i]
    }
  }
}

# combine data
cibersort_results_vafs <- inner_join(CIBERSORT_results, final_data_matrix, by = "Sample")

write.csv(cibersort_results_vafs, "~/Desktop/Majeti_Lab/Data/CIBERSORTx/MetaAML_clonality_results/master_data.csv", row.names=FALSE)


muts <- as.data.frame(unique(cibersort_results_vafs$Gene))


## perform variant allele frequency per-mutation regression with cell type abundances ####
# create list to populate results into
cell_type_vaf_list <- list()

z <- 1
# k <- 1
for(i in 1:nrow(muts)){
  print(i)
  
  gene <- as.character(muts[i,])
  sub <- subset(cibersort_results_vafs, cibersort_results_vafs$Gene == gene)
  n <- as.numeric(nrow(sub))
  
  if(n >= 5){
    
    for(j in 8:40){
      print(j)
      type <- as.character(colnames(sub[j]))
      
      n2 <- as.numeric(sum(sub[,j] != 0))
      
      if(n2 >= 5){
        
        # fit the data to a linear regression model
        lmf <- lm(sub$VAF ~ sub[[j]], data=sub)
        
        # calculate the slope of the line to determine if the relationship show resistance or sensitivity
        slp <-  as.numeric(coef(lmf)[2] )
        
        # extract the r-squared value
        lmr <- summary(lmf)$r.squared
        lmr <- round(lmr, 3)
        # extract the p-value
        lmp <- as.data.frame(summary(lmf)$coefficients[,4])
        lmp <- as.numeric(lmp[2,1])
        
        
        # find the delta in AUN and VAF for each case in order to filter later
        delta_vaf <- diff(range(sub$VAF, na.rm = T))
        delta_abundance <- diff(range(sub[[j]]))
        
        
        # store all of the relationships in a dataframe
        cell_type_vaf <- data.frame(matrix(NA, nrow = 1, ncol = 7))
        names(cell_type_vaf) <- c("Gene", "Cell_type", "Slope", "abundance_delta", "vaf_delta", "R_squared", "p_value")
        
        cell_type_vaf[1,1] <- gene
        cell_type_vaf[1,2] <- type
        cell_type_vaf[1,3] <- slp
        cell_type_vaf[1,4] <- delta_vaf
        cell_type_vaf[1,5] <- delta_abundance
        cell_type_vaf[1,6] <- lmr
        cell_type_vaf[1,7] <- lmp
        
        # Add each list in the loop to a list of lists
        z <- z + 1
        cell_type_vaf_list[[z]] <- cell_type_vaf
        
        # print the regression
        if(lmp < 0.05 & delta_vaf >= 0.25 & delta_abundance >= 0.25 & lmr > .25){
        # if(lmp < 0.05){
          # k <- k + 1
          # colnames(sub)[j] <- paste("cell_type_", k, sep = "")
          
          ggscatter(sub, x = "VAF", y = paste(type),
                    color = c("darkblue"), size = 5, alpha = 0.5,
                    font.label = list(color = "black", size = 9, vjust = 0.5),
                    title = paste(gene, " & ", type, sep = ""),
                    xlab = "VAF",
                    xlim = c(0,1),
                    ylab = "Cell abundance",
                    ylim = c(0,1)) +
            geom_smooth(method = "lm", color = "black", alpha = .25) +
            theme(plot.title = element_text(hjust = 0.5)) +
            stat_cor(
              aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
              label.x = 0.3,
              label.y = 1
            )
          
          ggsave(filename = paste("~/Desktop/Majeti_Lab/Data/CIBERSORTx/MetaAML_clonality_results/cell_type_percent_vaf_correlation_plots/",gene, "_", type, ".png", sep = ""), dpi = 150, width = 5, height = 5, units = "in")
        }
      }
    }
  }
}

MetaAML_cell_type_vaf_regression <- do.call(rbind, cell_type_vaf_list)

MetaAML_cell_type_vaf_regression$fdr <- p.adjust(MetaAML_cell_type_vaf_regression$p_value)

MetaAML_cell_type_vaf_regression$label <- paste(MetaAML_cell_type_vaf_regression$Gene, "_", MetaAML_cell_type_vaf_regression$Cell_type, sep = "")


write.csv(MetaAML_cell_type_vaf_regression, "~/Desktop/Majeti_Lab/Data/CIBERSORTx/MetaAML_clonality_results/MetaAML_cell_type_vaf_regression.csv",  row.names=FALSE)


# plot the results
  MetaAML_cell_type_vaf_regression <- subset(MetaAML_cell_type_vaf_regression, MetaAML_cell_type_vaf_regression$p_value <= 0.05 & MetaAML_cell_type_vaf_regression$vaf_delta >= 0.2 & MetaAML_cell_type_vaf_regression$abundance_delta >= 0.1 & MetaAML_cell_type_vaf_regression$R_squared >= .25)

  
  # for visualization, change the sign of the VAF delta based on the sign of the slope
  MetaAML_cell_type_vaf_regression$slope_sign <- "positive"
  
  for(i in 1:nrow(MetaAML_cell_type_vaf_regression)){
    if(MetaAML_cell_type_vaf_regression$Slope[i] < 0){
      MetaAML_cell_type_vaf_regression$vaf_delta[i] <- MetaAML_cell_type_vaf_regression$vaf_delta[i]*(-1)
      MetaAML_cell_type_vaf_regression$slope_sign[i] <- "negative"
      MetaAML_cell_type_vaf_regression$slope_sign[i] <- "negative"
      
    }
  }
  
mins <- as.numeric(min(MetaAML_cell_type_vaf_regression$vaf_delta) - .1)
maxs <- as.numeric(max(MetaAML_cell_type_vaf_regression$vaf_delta) + .1)
mids <- ((mins+maxs)/2)



ggplot(data = MetaAML_cell_type_vaf_regression, aes(x = MetaAML_cell_type_vaf_regression$Gene, y = MetaAML_cell_type_vaf_regression$Cell_type, fill = vaf_delta))+
  geom_tile(color = "white")+
  scale_fill_gradient2(low = "#08519c", high = "#a50f15", mid = "grey",
                       midpoint = mids, limit = c(mins,maxs), space = "Lab", 
                       name="VAF range\nRed (+ slope)\nBlue (- slope)") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                   size = 12, hjust = 1),
        panel.grid.major = element_line(size = 0.5, linetype = 'solid',
                                        colour = "#f0f0f0"))+
  xlab(label= "Mutated Gene") +
  ylab(label="Cell Type Signature/cluster") +
  labs(title = "VAF correlation with CIBERSORTx results\n(p < 0.05, vaf_delta > 0.2, abundance_delta > 0.2, R_squared > 0.25)") 

ggsave(filename = "~/Desktop/Majeti_Lab/Data/CIBERSORTx/MetaAML_clonality_results/VAF_effects_on_CIBERSORT_results.png", dpi = 300, width = 10, height = 7.5, units = "in")



 ## plot distribution results as a volcano plot
ggplot(MetaAML_cell_type_vaf_regression, aes(x=MetaAML_cell_type_vaf_regression$vaf_delta, y=MetaAML_cell_type_vaf_regression$abundance_delta, color = factor(slope_sign))) +
  geom_point(alpha = 0.5, size = 3) + 
  geom_label_repel(aes(label=label),hjust=0, vjust=1.5, size = 4) +
  scale_colour_manual(values = c("positive"= "#bf812d", "negative"="#35978f")) +
  theme(legend.position="none") +
  geom_vline(xintercept = 0,  linetype = "dashed", color = "grey") +
  # geom_vline(xintercept = 0, linetype = "dashed", color = "grey") +
  xlab(label= "VAF range") +
  ylab(label = "Cell type abundance difference") +
  # xlab(label= expression(Delta~"sgRNA dropout (mean(mut) - mean(not mut))")) +
  labs(title = "VAF correlation with CIBERSORTx results (MetaAML)") +
  theme(plot.title = element_text(color="black", size=20)) 

ggsave(filename = "~/Desktop/Majeti_Lab/Data/CIBERSORTx/MetaAML_clonality_results/VAF_effects_on_CIBERSORT_results_points.png", dpi = 300, width = 10, height = 7.5, units = "in")





## perform binary per-mutation comparison with cell-type abundance ####
# create list to populate results into
cell_type_binary_list <- list()
cibersort_results_vafs_sub <- select(cibersort_results_vafs, Sample, 8:40)

z <- 1
# k <- 1
for(i in 1:nrow(muts)){
  print(i)
  
  gene <- as.character(muts[i,])
  sub1 <- subset(cibersort_results_vafs_sub, cibersort_results_vafs_sub$Gene == gene)
  # sub1[,2:4] <- NULL
  sub1 <- distinct_all(sub1)
  sub1$Status <- 1
  
  sub2 <- setDT(cibersort_results_vafs_sub)[! Sample %chin% sub1$Sample] 
  # sub2[,2:4] <- NULL
  sub2 <- distinct_all(sub2)
  sub2$Status <- 0
  
  sub_final <- rbind(sub1, sub2)
  # sub_final <- subset(sub_final, select=c(17,1:16))
  
  n <- as.numeric(nrow(sub1))
  
  if(n >= 5){
    
    for(j in 3:34){
      print(j)
      type <- colnames(sub_final)[j]
      
      p_value <- as.numeric(t.test(sub_final[[j]]~sub_final$Status)$p.value)
      
      mean_mut <- subset(sub_final, sub_final$Status == 1)
      mean_mut <- mean(mean_mut[[j]])
      
      mean_wt <- subset(sub_final, sub_final$Status == 0)
      mean_wt <- mean(mean_wt[[j]])
      
      diff <- (mean_wt - mean_mut)
      
      # store all of the relationships in a dataframe
      cell_type_mutation <- data.frame(matrix(NA, nrow = 1, ncol = 4))
      names(cell_type_mutation) <- c("Gene", "Cell_type", "delta_mean_wt_minus_mut", "p_value")
      
      cell_type_mutation[1,1] <- gene
      cell_type_mutation[1,2] <- type
      cell_type_mutation[1,3] <- diff
      cell_type_mutation[1,4] <- p_value
      
      # Add each list in the loop to a list of lists
      z <- z + 1
      cell_type_binary_list[[z]] <- cell_type_mutation
      
      # print the regression
      if(p_value < 0.05){
        # k <- k + 1
        # colnames(sub_final)[j] <- paste("cell_type_", k, sep = "")
        
        ggviolin(sub_final, x = "Status", y = paste(type),
                  color = "Status", fill = "Status", add = "jitter", alpha = 0.5, palette = c("#08519c", "#a50f15"), bxp.errorbar = F, outlier.shape = NA) +
          stat_compare_means(method = "t.test", label.x = 1.4, label = "p.format") +
          labs(title = paste(gene, "and", type, "signature", sep = " ")) +
          ylab(label= paste("% ", type, "signature")) +
          scale_x_discrete(labels=c("0" = paste(gene, " WT"), "1" = paste(gene, " Mutated"))) +
          theme(plot.title = element_text(hjust = 0.5))
        
        # save cases where the mutation correlates with lower cell type abundance
        if(diff > 0){
          ggsave(filename = paste("~/Desktop/Majeti_Lab/Data/CIBERSORTx/MetaAML_clonality_results/cell_type_percent_mutation_correlation_plots/cell_type_decreased/",gene, "_", type, ".png", sep = ""), dpi = 150, width = 5, height = 5, units = "in")  
        }                                                                                                                                                               
        # save cases where the mutation correlates with lower cell type abundance                                                                                                                 
        if(diff < 0){
          ggsave(filename = paste("~/Desktop/Majeti_Lab/Data/CIBERSORTx/MetaAML_clonality_results/cell_type_percent_mutation_correlation_plots/cell_type_increased/",gene, "_", type, ".png", sep = ""), dpi = 150, width = 5, height = 5, units = "in")
        }                       
      }
    }
  }
}



MetaAML_cell_type_binary_list <- do.call(rbind, cell_type_binary_list)

MetaAML_cell_type_binary_list$fdr <- p.adjust(MetaAML_cell_type_binary_list$p_value)

write.csv(MetaAML_cell_type_binary_list, "~/Desktop/Majeti_Lab/Data/CIBERSORTx/MetaAML_clonality_results/MetaAML_cell_type_binary_list.csv",  row.names=FALSE)


# plot the results
MetaAML_cell_type_binary_list <- subset(MetaAML_cell_type_binary_list, MetaAML_cell_type_binary_list$fdr <= 0.1)
mins <- as.numeric(min(MetaAML_cell_type_binary_list$delta_mean_wt_minus_mut) - .1)
maxs <- as.numeric(max(MetaAML_cell_type_binary_list$delta_mean_wt_minus_mut) + .1)
mids <- ((mins+maxs)/2)



ggplot(data = MetaAML_cell_type_binary_list, aes(x = MetaAML_cell_type_binary_list$Gene, y = MetaAML_cell_type_binary_list$Cell_type, fill = delta_mean_wt_minus_mut))+
  geom_tile(color = "white")+
  scale_fill_gradient2(low = "#08519c", high = "#a50f15", mid = "grey",
                       midpoint = mids, limit = c(mins,maxs), space = "Lab", 
                       name="Difference in mean\ncell type % (wt - mut)") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                   size = 12, hjust = 1),
        panel.grid.major = element_line(size = 0.5, linetype = 'solid',
                                        colour = "#f0f0f0"))+
  xlab(label= "Mutated Gene") +
  ylab(label="Cell Type") +
  labs(title = "Mutation correlation with CIBERSORT results (q < 0.1)") 

ggsave(filename = "~/Desktop/Majeti_Lab/Data/CIBERSORTx/MetaAML_clonality_results/Mutation_effects_on_CIBERSORT_results.png", dpi = 300, width = 10, height = 6, units = "in")


